### DISCUSSION TITLE
How to get the raw request body in action or loader ( for stripe integration) ?

### DISCUSSION THREAD

**Discussion by marvinwu** [remix-run/remix#1978] [Q&A] [answered]
For stripe webhook integration, it requires to pass in the request body to [verify the event](https://github.com/stripe/stripe-node#webhook-signing)

```
const event = stripe.webhooks.constructEvent(
  webhookRawBody,
  webhookStripeSignatureHeader,
  webhookSecret
);
```

in express server we could do it via 

```
app.post('/webhook', express.raw({type: 'application/json'}), (req, res) => {
```

but how can we do it in remix?

**Comment by sergiodxa** [comment, marked as answer]
You can call request.text() to get it as a string or request.json() if it’s a JSON and you need it as a JS object.

**Reply by marvinwu** [reply]
Thx Sergio for the tips, seems the  stripe.webhooks.constructEvent need to be feed with request raw buffer body, so neither request.json() and request.text() is working, we have to disable request parsing to make stripe happy, need something similar to this in nextjs

```
export const config = {
  api: {
    bodyParser: false,
  },
};
```

**Reply by sergiodxa** [reply]
Ahh I think the issue is `stripe.webhooks.constructEvent` probably expects the request body from a Node request, but Remix uses Fetch standard Request objects.

If that's the case you may need to move your webhook outside Remix to Express or Vercel API routes so you can access that request object.

Something like `bodyParser: false` will not work in Remix because Remix doesn't use bodyParser, you parse the body when you do `request.json()` or `request.text()` or `request.formData()`.

**Reply by marvinwu** [reply]
@sergiodxa thanks much,  appreciate the help. probably will try move the hook outside of remix then

**Reply by kiliman** [reply]
Hmm... what error are you getting? I was able to use `request.text()` to pass to stripe API.
```ts
export const action: ActionFunction = async ({ request }) => {
  const payload = await request.text()
  console.log('payload', payload)
  const sig = request.headers.get('stripe-signature')
  let event
  try {
    event = stripe.webhooks.constructEvent(payload, sig, endpointSecret)
  } catch (err: any) {
    console.log(err)
    throw json(
      {
        errors: [
          {
            message: err.message,
          },
        ],
      },
      400,
    )
  }
  console.log(event)
  return new Response(null, { status: 200 })
}
```

**Reply by marvinwu** [reply]
@kiliman thanks much for helping, i was keep getting the expected signature not found error, i am started to think probably it is a vercel bug( my remix build is on vercel adapter), which deployment option are you using ?

```
StripeSignatureVerificationError: No signatures found matching the expected signature for payload. Are you passing the raw request body you received from Stripe? https://github.com/stripe/stripe-node#webhook-signing
```

**Reply by kiliman** [reply]
That was on Netlify (I was helping somebody else with Stripe integration). So yes, it's possible it could be a Vercel issue.

**Reply by marvinwu** [reply]
@kiliman thx much!  request.text() works like a charm, i figured out why it doesnt work earlier - was forwarding the webhook locally and didn't aware  it needs a different signing secrets. I was using the wrong signing secrets all along

@sergiodxa you are right in the beginning, request.text() does work!
